internal/generator/backend/template/zkpschemes/groth16/mpcsetup/utils.go.tmpl

```diff
@@ -84,70 +66,20 @@ func scaleG2InPlace(A []curve.G2Affine, a []fr.Element) {
 	})
 }
 
-// Check e(a₁, a₂) = e(b₁, b₂)
-func sameRatio(a1, b1 curve.G1Affine, a2, b2 curve.G2Affine) bool {
-	if !a1.IsInSubGroup() || !b1.IsInSubGroup() || !a2.IsInSubGroup() || !b2.IsInSubGroup() {
-		panic("invalid point not in subgroup")
-	}
-	var na2 curve.G2Affine
-	na2.Neg(&a2)
-	res, err := curve.PairingCheck(
-		[]curve.G1Affine{a1, b1},
-		[]curve.G2Affine{na2, b2})
-	if err != nil {
-		panic(err)
-	}
-	return res
-}
-
-// returns a = ∑ rᵢAᵢ, b = ∑ rᵢBᵢ
-func merge(A, B []curve.G1Affine) (a, b curve.G1Affine) {
-	nc := runtime.NumCPU()
-	r := make([]fr.Element, len(A))
-	for i := 0; i < len(A); i++ {
-		r[i].SetRandom()
-	}
-	a.MultiExp(A, r, ecc.MultiExpConfig{NbTasks: nc / 2})
-	b.MultiExp(B, r, ecc.MultiExpConfig{NbTasks: nc / 2})
-	return
-}
-
-// L1 = ∑ rᵢAᵢ, L2 = ∑ rᵢAᵢ₊₁ in G1
-func linearCombinationG1(A []curve.G1Affine) (L1, L2 curve.G1Affine) {
-	nc := runtime.NumCPU()
-	n := len(A)
-	r := make([]fr.Element, n-1)
-	for i := 0; i < n-1; i++ {
-		r[i].SetRandom()
-	}
-	L1.MultiExp(A[:n-1], r, ecc.MultiExpConfig{NbTasks: nc / 2})
-	L2.MultiExp(A[1:], r, ecc.MultiExpConfig{NbTasks: nc / 2})
-	return
-}
-
-// L1 = ∑ rᵢAᵢ, L2 = ∑ rᵢAᵢ₊₁ in G2
-func linearCombinationG2(A []curve.G2Affine) (L1, L2 curve.G2Affine) {
-	nc := runtime.NumCPU()
-	n := len(A)
-	r := make([]fr.Element, n-1)
-	for i := 0; i < n-1; i++ {
-		r[i].SetRandom()
+func areInSubGroupG1(s []curve.G1Affine) bool {
+	for i := range s {
+		if !s[i].IsInSubGroup() {
```

Comment:
that's not cheap, we don't have a batch subgroup check yet (@yelhousni wanted to add one) but shouldn't we parallelize since we assume the nominal case is everything is in subgroup ?